pallets/marketplace/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+				Error::<T>::NotTheTokenOwner
+			);
+
+			Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
+				let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;
+
+				token_info.price = new_price;
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::TokenPriceUpdated(sender, class_id, instance_id, new_price));
+
+			Ok(())
+		}
+
+		/// Lists the token on Marketplace
+		/// freezes the NFT from transfers
+		/// and other modifications
+		///
+		/// Parameters:
+		/// - `class_id`: The identifier of a non-fungible token class
+		/// - `instance_id`: The instance identifier of a class
+		#[pallet::weight(<T as Config>::WeightInfo::list())]
+		#[transactional]
+		pub fn list(origin: OriginFor<T>, class_id: T::NftClassId, instance_id: T::NftInstanceId) -> DispatchResult {
+			let sender = ensure_signed(origin.clone())?;
+
+			let class_type = pallet_nft::Classes::<T>::get(class_id)
+				.map(|c| c.class_type)
+				.ok_or(pallet_nft::Error::<T>::ClassUnknown)?;
+
+			ensure!(
+				!Tokens::<T>::contains_key(class_id, instance_id),
+				Error::<T>::AlreadyListed
+			);
+			ensure!(class_type == ClassType::Marketplace, Error::<T>::UnsupportedClassType);
+
+			let royalty = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.royalty)
+				.ok_or(pallet_nft::Error::<T>::RoyaltyNotSet)?;
+			let author = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.author)
+				.ok_or(pallet_nft::Error::<T>::AuthorNotSet)?;
+
+			// Only token owner can list
+			ensure!(
+				pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into()) == Some(sender.clone()),
+				Error::<T>::NotTheTokenOwner
+			);
+
+			Tokens::<T>::insert(class_id, instance_id, TokenInfo { price: None });
+
+			pallet_uniques::Pallet::<T>::freeze(origin.clone(), class_id.into(), instance_id.into())?;
+
+			Self::deposit_event(Event::TokenListed(sender, class_id, instance_id, author, royalty));
+
+			Ok(())
+		}
+
+		/// Unlists the token from Marketplace
+		/// unfreezes the NFT from transfers
+		/// and other modifications
+		///
+		/// Parameters:
+		/// - `class_id`: The identifier of a non-fungible token class
+		/// - `instance_id`: The instance identifier of a class
+		#[pallet::weight(<T as Config>::WeightInfo::unlist())]
+		#[transactional]
+		pub fn unlist(origin: OriginFor<T>, class_id: T::NftClassId, instance_id: T::NftInstanceId) -> DispatchResult {
+			let sender = ensure_signed(origin.clone())?;
+
+			ensure!(Tokens::<T>::contains_key(class_id, instance_id), Error::<T>::NotListed);
+
+			ensure!(
+				pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into()) == Some(sender),
+				Error::<T>::NotTheTokenOwner
+			);
+
+			let token_id = (class_id, instance_id);
+
+			Tokens::<T>::remove(class_id, instance_id);
+			Offers::<T>::remove_prefix(token_id, None);
+
+			let owner = pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into())
```

Comment:
let's move this check before all storage changes.  

Even if it is transactional, it is better to check first and then make modifications.